File: backend/app/schemas/organization.py
"""Validated API contracts for P006 company structure management."""

from uuid import UUID

from pydantic import BaseModel, Field


class AddressFields(BaseModel):
    address_line1: str | None = Field(default=None, max_length=255)
    address_line2: str | None = Field(default=None, max_length=255)
    city: str | None = Field(default=None, max_length=128)
    state: str | None = Field(default=None, max_length=128)
    postal_code: str | None = Field(default=None, max_length=32)
    country: str | None = Field(default=None, max_length=128)


class CompanyUpdateRequest(AddressFields):
    legal_name: str | None = Field(default=None, max_length=255)
    display_name: str | None = Field(default=None, max_length=255)
    business_type: str = Field(default="proprietorship", min_length=2, max_length=64)
    gst_status: str | None = Field(default=None, max_length=32)
    gstin: str | None = Field(default=None, max_length=32)
    email: str | None = Field(default=None, max_length=320)
    phone: str | None = Field(default=None, max_length=32)
    logo_url: str | None = Field(default=None, max_length=1024)


class CompanyResponse(AddressFields):
    id: UUID
    business_name: str
    legal_name: str | None
    display_name: str | None
    business_type: str
    status: str
    gst_status: str | None
    gstin: str | None
    email: str | None
    phone: str | None
    logo_url: str | None
    setup_progress: int


class BranchInput(AddressFields):
    name: str = Field(min_length=1, max_length=255)
    code: str | None = Field(default=None, max_length=64)
    email: str | None = Field(default=None, max_length=320)
    phone: str | None = Field(default=None, max_length=32)


class BranchResponse(BranchInput):
    id: UUID
    is_active: bool


class WarehouseInput(AddressFields):
    name: str = Field(min_length=1, max_length=255)
    code: str | None = Field(default=None, max_length=64)
    branch_id: UUID | None = None
    contact_person: str | None = Field(default=None, max_length=255)
    contact_number: str | None = Field(default=None, max_length=32)


class WarehouseResponse(WarehouseInput):
    id: UUID
    is_active: bool


class ActiveStatusRequest(BaseModel):
    is_active: bool

File: frontend/services/organization.ts
import type { AccessibleCompany } from "./auth";

const apiBaseUrl = import.meta.env.VITE_API_BASE_URL ?? "http://localhost:8000";

export type AddressFields = {
  address_line1: string | null;
  address_line2: string | null;
  city: string | null;
  state: string | null;
  postal_code: string | null;
  country: string | null;
};

export type CompanyProfile = AddressFields & {
  id: string;
  business_name: string;
  legal_name: string | null;
  display_name: string | null;
  business_type: string;
  status: "active" | "suspended" | "closed";
  gst_status: string | null;
  gstin: string | null;
  email: string | null;
  phone: string | null;
  logo_url: string | null;
  setup_progress: number;
};

export type Branch = AddressFields & {
  id: string;
  name: string;
  code: string | null;
  email: string | null;
  phone: string | null;
  is_active: boolean;
};

export type Warehouse = AddressFields & {
  id: string;
  name: string;
  code: string | null;
  branch_id: string | null;
  contact_person: string | null;
  contact_number: string | null;
  is_active: boolean;
};

export type CompanyUpdate = Omit<CompanyProfile, "id" | "business_name" | "status" | "setup_progress">;
export type BranchInput = Omit<Branch, "id" | "is_active">;
export type WarehouseInput = Omit<Warehouse, "id" | "is_active">;

async function request<T>(path: string, companyId: string, method = "GET", body?: object): Promise<T> {
  const response = await fetch(`${apiBaseUrl}${path}`, {
    method,
    credentials: "include",
    headers: { "X-Company-ID": companyId, ...(body ? { "Content-Type": "application/json" } : {}) },
    body: body ? JSON.stringify(body) : undefined,
  });
  const payload = await response.json().catch(() => ({}));
  if (!response.ok) throw new Error(payload.detail ?? "Organization data could not be updated.");
  return payload as T;
}

export const organizationApi = {
  company: (companyId: string) => request<CompanyProfile>("/organization/company", companyId),
  updateCompany: (companyId: string, payload: CompanyUpdate) => request<CompanyProfile>("/organization/company", companyId, "PUT", payload),
  branches: (companyId: string) => request<Branch[]>("/organization/branches", companyId),
  createBranch: (companyId: string, payload: BranchInput) => request<Branch>("/organization/branches", companyId, "POST", payload),
  updateBranch: (companyId: string, branchId: string, payload: BranchInput) => request<Branch>(`/organization/branches/${branchId}`, companyId, "PUT", payload),
  setBranchStatus: (companyId: string, branchId: string, is_active: boolean) => request<Branch>(`/organization/branches/${branchId}/status`, companyId, "PATCH", { is_active }),
  warehouses: (companyId: string) => request<Warehouse[]>("/organization/warehouses", companyId),
  createWarehouse: (companyId: string, payload: WarehouseInput) => request<Warehouse>("/organization/warehouses", companyId, "POST", payload),
  updateWarehouse: (companyId: string, warehouseId: string, payload: WarehouseInput) => request<Warehouse>(`/organization/warehouses/${warehouseId}`, companyId, "PUT", payload),
  setWarehouseStatus: (companyId: string, warehouseId: string, is_active: boolean) => request<Warehouse>(`/organization/warehouses/${warehouseId}/status`, companyId, "PATCH", { is_active }),
};

export function selectedCompanyName(company?: AccessibleCompany): string {
  return company?.business_name ?? "your company";
}
